### PR TITLE
lock workspace

### PR DESCRIPTION
**Pull request by tedsharpe**

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -163,7 +167,7 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
                   gcsDAO.deleteBucket(userInfo.userId, workspaceRequest.namespace, bucketName)
                   RequestComplete(StatusCodes.Forbidden, s"Unable to create groups for ${workspaceRequest.namespace}/${workspaceRequest.name}: "+err.getMessage)
                 case Success(_) =>
-                  val workspace = Workspace(workspaceRequest.namespace, workspaceRequest.name,bucketName,DateTime.now,userInfo.userId,workspaceRequest.attributes)
+                  val workspace = Workspace(workspaceRequest.namespace, workspaceRequest.name,bucketName,DateTime.now,userInfo.userId,false,workspaceRequest.attributes)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Prefer including parameter names with booleans, like `isLocked = false`

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 2]
```diff
@@ -523,4 +529,120 @@ class WorkspaceApiServiceSpec extends FlatSpec with HttpService with ScalatestRo
...
+      check {
+        assertResult(StatusCodes.NotFound) { status }
+      }
+    Put(s"/workspaces/${testData.workspace.namespace}/${testData.workspace.name}/unlock") ~>
+      addOpenAmCookie("no-access") ~>
+      sealRoute(services.workspaceRoutes) ~>
+      check {
+        assertResult(StatusCodes.NotFound) { status }
+      }
+  }
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: it should "allow a reader to copy a locked workspace" ?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -285,21 +283,49 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
...
     }
 
   def updateACL(workspaceName: WorkspaceName, aclUpdates: Seq[WorkspaceACLUpdate]): PerRequestMessage =
     dataSource inTransaction { txn =>
-      withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevel.Owner, txn) { workspaceContext =>
-        val updateErrors = gcsDAO.updateACL(workspaceContext.bucketName, workspaceName, aclUpdates)
-        updateErrors.size match {
-          case 0 => RequestComplete(StatusCodes.OK)
-          case _ => RequestComplete(StatusCodes.Conflict, updateErrors)
+      withWorkspaceContext(workspaceName, txn) { workspaceContext =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Why the change from `withWorkspaceContextAndPermissions` to `withWorkspaceContext` and `asOwner` here, but the opposite refactoring above (line 192)?

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Ah, because `asOwner` bypasses the locking code.  I'd like this to be more explicit, but I don't have any ideas right now besides unwieldy method names.

Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Because someone just failed to combine the two on line 192 so I fixed it for them, but this one is one of the special ones that owners can do even when there's a lock.  I need a better name for asOwner that makes this clearer.  requireOwnerIgnoreLock, maybe.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 2 above)
Comment (tedsharpe) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Don't quite get it -- you feel we need a special test for this particular example of reading from a locked workspace?

Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Just a suggestion. Without scrutinizing I don't know if copying a workspace requires WRITE or something (not sure why it would).

Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Copying requires READ on the source, WRITE on the dest.

But now I'm wondering: Should the new copied workspace be in the locked state?

Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: No, it shouldn't.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 4]
```diff
@@ -523,4 +529,120 @@ class WorkspaceApiServiceSpec extends FlatSpec with HttpService with ScalatestRo
 
   // End ACL-restriction Tests
 
+  // Workspace Locking
+  it should "allow an owner to lock the workspace" in withEmptyWorkspaceApiServices { services =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Should it `allow the owner to lock an already locked workspace` ?  I don't know what the desired behavior is for that, but we should test it once we have that answer.

**Comment by jmthibault79** [comment]
Done with my review.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 4 above)
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: PUT should be idempotent so I'd say locking a locked workspace is OK.

**Comment by helgridly** [comment]
Also done commenting.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 4 above)
Comment (tedsharpe) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: As, indeed, it is with existing code.  I'll add a test.

**Comment by tedsharpe** [comment]
I believe that all comments have been addressed, and this PR is based on master and squashed.  If a re-review turns up no new issues, I believe we're ready for a merge.

**Comment by helgridly** [comment]
Next time, can you check in your post-comment commit unsquashed? That way reviewers can look at that single commit and just see the delta.

**Comment by tedsharpe** [comment]
This is not ready to merge.  Need to prevent propagation of lock status.

**Comment by tedsharpe** [comment]
Ready to merge (after a squash), IMHO.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAO.scala [hunk 5]
```diff
@@ -56,4 +56,18 @@ class GraphWorkspaceDAO extends WorkspaceDAO with GraphDAO {
...
+    workspaceContext.workspaceVertex.getProperty(LOCK_PROPERTY)
+  }
+
+  override def lock(workspaceContext: WorkspaceContext): Unit = {
+    workspaceContext.workspaceVertex.setProperty(LOCK_PROPERTY,true)
+  }
+
+  override def unlock(workspaceContext: WorkspaceContext): Unit = {
+    workspaceContext.workspaceVertex.setProperty(LOCK_PROPERTY,false)
+  }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAO.scala]: why do we need special functions for this? shouldn't we just change the property on the Workspace and call save?

**Comment by dvoet** [comment]
I am good with this with the exception of the custom functions to set the lock vertex property. I think that circumvents all of Hussein's hard work to make vertex manipulation centralized and generic.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAO.scala (hunk 5 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAO.scala]: Yes, but in order for the field to show up on the Context you'd have to
pull it out of the vertex manually like we do with bucketName.

Workspaces are a bit of a special case in that they have multiple scala
representations, none of which fully reflect the vertex contents. I'm okay
with this.
On Sep 4, 2015 9:13 AM, "dvoet" <notifications@github.com> wrote:

In
src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAO.scala
<https://github.com/broadinstitute/rawls/pull/113#discussion_r38749756>:

> @@ -56,4 +56,18 @@ class GraphWorkspaceDAO extends WorkspaceDAO with GraphDAO {
>        case None => false
>      }
>    }
> +
> +  val LOCK_PROPERTY = "isLocked"
> +
> +  override def isLocked(workspaceContext: WorkspaceContext): Boolean = {
> +    workspaceContext.workspaceVertex.getProperty(LOCK_PROPERTY)
> +  }
> +
> +  override def lock(workspaceContext: WorkspaceContext): Unit = {
> +    workspaceContext.workspaceVertex.setProperty(LOCK_PROPERTY,true)
> +  }
> +
> +  override def unlock(workspaceContext: WorkspaceContext): Unit = {
> +    workspaceContext.workspaceVertex.setProperty(LOCK_PROPERTY,false)
> +  }

why do we need special functions for this? shouldn't we just change the
property on the Workspace and call save?

—
Reply to this email directly or view it on GitHub
<https://github.com/broadinstitute/rawls/pull/113/files#r38749756>.

**Comment by helgridly** [comment]
:+1: when build goes green.
